# Quality Assessment
Now it's your turn. Follow the steps on the platform and use what you've learnt to see how reliable the data is.

In [ ]:
import pandas as pd

Load our cleaned DataFrames

In [ ]:
# orders_cl.csv

# products_cl.csv

# orderlines_cl.csv

In [ ]:
from google.colab import files
import pandas as pd
import io

In [ ]:
# 1. Upload the first file (Orderlines_clean)
print("Select the ORDERLINES_clean file:")
uploaded_1 = files.upload()
filename1 = list(uploaded_1.keys())[0]
orderlines_clean = pd.read_csv(io.BytesIO(uploaded_1[filename1]))

Select the ORDERLINES_clean file:


Saving orderlines_cleaned.csv to orderlines_cleaned (1).csv


In [ ]:
# 2. Upload the second file (Orders_clean)
print("Select the ORDERS_clean file:")
uploaded_1 = files.upload()
filename1 = list(uploaded_1.keys())[0]
orders_clean = pd.read_csv(io.BytesIO(uploaded_1[filename1]))

Select the ORDERS_clean file:


Saving orders_cleaned.csv to orders_cleaned (2).csv


In [ ]:
# 3. Upload the third file (Products_clean)
print("Select the PRODUCTS_clean file:")
uploaded_1 = files.upload()
filename1 = list(uploaded_1.keys())[0]
products_clean = pd.read_csv(io.BytesIO(uploaded_1[filename1]))

Select the PRODUCTS_clean file:


Saving products_cleaned.csv to products_cleaned (1).csv


## 1.&nbsp; Define Pandas display format

In [ ]:
#make a copy
orderlines_cl_df = orderlines_clean.copy()
orders_cl_df = orders_clean.copy()
products_cl_df = products_clean.copy()

In [ ]:
orderlines_cl_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        293983 non-null  int64  
 1   id                293983 non-null  int64  
 2   id_order          293983 non-null  int64  
 3   product_id        293983 non-null  int64  
 4   product_quantity  293983 non-null  int64  
 5   sku               293983 non-null  object 
 6   unit_price        293983 non-null  float64
 7   date              293983 non-null  object 
 8   total_revenue     293983 non-null  float64
dtypes: float64(2), int64(5), object(2)
memory usage: 20.2+ MB


In [ ]:
orders_cl_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   order_id      226909 non-null  int64 
 1   created_date  226909 non-null  object
 2   total_paid    226909 non-null  object
 3   state         226909 non-null  object
dtypes: int64(1), object(3)
memory usage: 6.9+ MB


In [ ]:
products_cl_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   sku                  19326 non-null  object 
 1   name                 19326 non-null  object 
 2   desc                 19326 non-null  object 
 3   price                19326 non-null  object 
 4   promo_price          19326 non-null  object 
 5   in_stock             19326 non-null  bool   
 6   type                 19276 non-null  object 
 7   price_cleaned        19326 non-null  float64
 8   promo_price_cleaned  19326 non-null  float64
 9   price_ratio          19326 non-null  float64
dtypes: bool(1), float64(3), object(6)
memory usage: 1.3+ MB


In [ ]:
# display columns of products_cl_df
print(products_cl_df.columns.tolist())

['sku', 'name', 'desc', 'price', 'promo_price', 'in_stock', 'type', 'price_cleaned', 'promo_price_cleaned', 'price_ratio']


In [ ]:
# display columns of orders_cl_df
print(orders_cl_df.columns.tolist())

['order_id', 'created_date', 'total_paid', 'state']


In [ ]:
# display columns of orderlines_cl_df
print(orderlines_cl_df.columns.tolist())

['Unnamed: 0', 'id', 'id_order', 'product_id', 'product_quantity', 'sku', 'unit_price', 'date', 'total_revenue']


In [ ]:
Step 1.1 Displaying Duplicates

In [ ]:
cols = ['sku', 'name', 'desc', 'price', 'promo_price', 'in_stock', 'type', 'price_cleaned', 'promo_price_cleaned', 'price_ratio']
print(products_cl_df.duplicated(subset=cols).sum())

8746


In [ ]:
# keep=False ensures all occurrences (including the first one) are shown
duplicates_view = products_cl_df[products_cl_df.duplicated(subset=cols, keep=False)]

# Sorting helps you see identical rows grouped together
duplicates_view.sort_values(by='sku')

,sku,name,desc,price,promo_price,in_stock,type,price_cleaned,promo_price_cleaned,price_ratio
1902,AII0013,"aiino case macbook air 13 ""transparent",macbook air 13-inch casing with matte finish.,39.99,179.927,False,13835403,39.99,17.99,0.4499
1903,AII0013,"aiino case macbook air 13 ""transparent",macbook air 13-inch casing with matte finish.,39.99,179.927,False,13835403,39.99,17.99,0.4499
107,APP0390,"apple macbook pro 133 ""core i5 25ghz | 4gb ram...",macbook pro laptop 133 inches (md101y / a).,1199,11.455.917,False,1282,1199.00,1145.59,0.9555
106,APP0390,"apple macbook pro 133 ""core i5 25ghz | 4gb ram...",macbook pro laptop 133 inches (md101y / a).,1199,11.455.917,False,1282,1199.00,1145.59,0.9555
105,APP0390,"apple macbook pro 133 ""core i5 25ghz | 4gb ram...",macbook pro laptop 133 inches (md101y / a).,1199,11.455.917,False,1282,1199.00,1145.59,0.9555
...,...,...,...,...,...,...,...,...,...,...
12714,WDT0311-A,open - wd my cloud ex2 ultra mac and pc server...,2-bay nas 1gb ram for mac and pc,169.99,1.325.138,True,11935397,169.99,132.51,0.7795
12715,WDT0311-A,open - wd my cloud ex2 ultra mac and pc server...,2-bay nas 1gb ram for mac and pc,169.99,1.325.138,True,11935397,169.99,132.51,0.7795
12716,WDT0311-A,open - wd my cloud ex2 ultra mac and pc server...,2-bay nas 1gb ram for mac and pc,169.99,1.325.138,True,11935397,169.99,132.51,0.7795
12723,WDT0317,wd my cloud ex2 ultra mac and pc server nas,2-bay nas 1gb ram for mac and pc,169.99,1.391.899,False,11935397,169.99,139.19,0.8188


Step 1.2 Deleting duplicates

1.2.1 Basic Removal (All Columns)

In [ ]:
#code: products_cl_df = products_cl_df.drop_duplicates()

In [ ]:
#display duplicates in each column
#code: products_cl_df.apply(lambda x: x.duplicated().sum())

1.2.2 Removal based on your Specific List

*   If you want to delete rows that match in your specific list of columns (even if other columns like an id or timestamp differ), use the subset parameter:



In [ ]:
cols = ['sku', 'name', 'desc', 'price', 'promo_price', 'in_stock', 'type', 'price_cleaned', 'promo_price_cleaned', 'price_ratio']

# Keep the first occurrence and delete the rest
products_cl_df = products_cl_df.drop_duplicates(subset=cols, keep='first')

1.2.3 Clean up the Index

*   After deleting rows, your index (the row numbers) will have gaps (e.g., 0, 1, 3, 4...). Use ignore_index=True to reset it automatically:

In [ ]:
products_cl_df = products_cl_df.drop_duplicates(subset=cols, ignore_index=True)

Comparing the row count before and after, you can use the .shape attribute, which returns a tuple of (rows, columns).


In [ ]:
# 1. Store the original count
original_count = products_cl_df.shape[0]

# 2. Drop the duplicates
cols = ['sku', 'name', 'desc', 'price', 'promo_price', 'in_stock', 'type', 'price_cleaned', 'promo_price_cleaned', 'price_ratio']
products_cl_df = products_cl_df.drop_duplicates(subset=cols, ignore_index=True)

# 3. Store the new count
new_count = products_cl_df.shape[0]

# 4. Display the results
print(f"Rows before: {original_count}")
print(f"Rows after:  {new_count}")
print(f"Removed:     {original_count - new_count}")

Rows before: 10580
Rows after:  10580
Removed:     0


In [ ]:
# 1. Store the original count
original_count = orders_cl_df.shape[0]

# 2. Drop the duplicates
cols = ['order_id', 'created_date', 'total_paid', 'state']
orders_cl_df = orders_cl_df.drop_duplicates(subset=cols, ignore_index=True)

# 3. Store the new count
new_count = orders_cl_df.shape[0]

# 4. Display the results
print(f"Rows before: {original_count}")
print(f"Rows after:  {new_count}")
print(f"Removed:     {original_count - new_count}")

Rows before: 226909
Rows after:  226909
Removed:     0


## 2.&nbsp; Exclude unwanted orders

In [ ]:
# your code here
orderlines_cl_df.sample(5)

,Unnamed: 0,id,id_order,product_id,product_quantity,sku,unit_price,date,total_revenue
20304,20304,1165492,317486,0,1,MIC0036,45.59,2017-01-30 13:43:24,45.59
138418,138418,1391344,419055,0,1,LGE0044,578.99,2017-10-28 19:33:08,578.99
92339,92339,1295848,378945,0,1,OWC0008,39.99,2017-07-18 11:58:26,39.99
126966,126966,1371059,409074,0,1,ELG0034,37.99,2017-10-05 22:01:21,37.99
179973,179973,1466328,451299,0,1,APP0666,319.99,2017-11-29 04:44:10,319.99


1. Exclude Unwanted Data (Cleaning)

Based on your sample, here is what you should remove or fix immediately:

*   Drop Unnamed: 0: This is a redundant index column created during a CSV export. It takes up memory and provides no value.
*   Handle product_quantity == 0: In your sample, several rows show a quantity of 0 but have a total_revenue. This is a data anomaly. Decide if these are cancelled orders or if the quantity should actually be 1. Usually, rows with 0 quantity should be excluded from revenue analysis.
*   Filter Outliers: Check for unit_price values of 0.00 or extremely high values (like the 2732.00 entry) to ensure they aren't test entries or "dummy" products.
*   Remove Duplicates: Ensure that the combination of id_order and sku is unique, unless a customer can add the same item twice as separate line items.


In [ ]:
# 1. Exclude Unwanted Data (Cleaning)

# Drop redundant index column
orderlines_cl_df = orderlines_cl_df.drop(columns=['Unnamed: 0'])
#orderlines_cl_df = orderlines_cl_df.iloc[:, 1:]

In [ ]:
# Handle product_quantity: Keep only rows where quantity > 0
orderlines_cl_df = orderlines_cl_df[orderlines_cl_df['product_quantity'] > 0]

In [ ]:
# Filter Outliers: Remove rows with unit_price of 0 (usually tests/freebies)
orderlines_cl_df = orderlines_cl_df[orderlines_cl_df['unit_price'] > 0]

In [ ]:
# Remove Duplicates: Drop identical line items within the same order
orderlines_cl_df = orderlines_cl_df.drop_duplicates(subset=['id_order', 'sku'], keep='first')

In [ ]:
orderlines_cl_df.sample(5)

,id,id_order,product_id,product_quantity,sku,unit_price,date,total_revenue
96867,1304317,382908,0,1,APP0407,84.99,2017-07-27 17:01:58,84.99
137279,1389421,418082,0,1,BEL0240,29.99,2017-10-26 13:37:46,29.99
148823,1412782,428054,0,1,ELA0045,19.99,2017-11-17 20:14:28,19.99
293727,1649730,527179,0,2,APP0698,9.99,2018-03-14 12:17:20,19.98
31727,1187487,327251,0,1,SAN0137,53.99,2017-02-21 17:41:44,53.99


2. Optimization & Feature Engineering

To make this data "smarter" for analysis, apply these optimizations:

Convert date to Datetime: Transform the date string into a proper

*   Convert date to Datetime: Transform the date string into a proper Pandas datetime object. This allows you to extract Hour of Day, Day of Week, or Month, which is critical for identifying peak shopping times.

*   Extract SKU Prefixes: Your SKUs follow a pattern (app, sat, wac, cru). Create a new category column by extracting the first three letters. This lets you analyze which product lines (e.g., "Apple" vs. "Satellite") perform best.

*   Revenue Validation: Create a calculated column: calc_revenue = product_quantity * unit_price. Compare this to your total_revenue. If they don't match, you may have hidden discounts or taxes that need to be accounted for.

*   Downcast Data Types: Change id and id_order to integers and unit_price to float32 to reduce the memory footprint if your dataset is large.


In [ ]:
# 2. Optimization & Feature Engineering

# Convert date to Datetime
orderlines_cl_df['date'] = pd.to_datetime(orderlines_cl_df['date'])

In [ ]:
# Extract SKU Prefixes as category (e.g., 'app', 'sat')
orderlines_cl_df['category'] = orderlines_cl_df['sku'].str[:3]

In [ ]:
# Revenue Validation
orderlines_cl_df['calc_revenue'] = orderlines_cl_df['product_quantity'] * orderlines_cl_df['unit_price']

In [ ]:
# Downcast Data Types to save memory
orderlines_cl_df['id'] = pd.to_numeric(orderlines_cl_df['id'], downcast='integer')
orderlines_cl_df['id_order'] = pd.to_numeric(orderlines_cl_df['id_order'], downcast='integer')
orderlines_cl_df['unit_price'] = pd.to_numeric(orderlines_cl_df['unit_price'], downcast='float')

In [ ]:
# Optional: Check if calc_revenue matches total_revenue
#orderlines_cl_df['revenue_diff'] = orderlines_cl_df['calc_revenue'] - orderlines_cl_df['total_revenue']

In [ ]:
orderlines_cl_df.sample(5)

,id,id_order,product_id,product_quantity,sku,unit_price,date,total_revenue,category,calc_revenue
288455,1642415,524183,0,1,SAT0054,74.99,2018-03-09 21:40:44,74.99,SAT,74.99
20303,1165491,317485,0,1,SYN0147,315.39,2017-01-30 13:42:13,315.39,SYN,315.39
248504,1577871,497451,0,1,SAT0053,64.99,2018-01-24 04:48:01,64.99,SAT,64.99
6819,1137465,305826,0,1,OLL0020,89.99,2017-01-09 16:08:28,89.99,OLL,89.99
286112,1638765,522592,0,1,TRK0009,26.99,2018-03-07 11:38:37,26.99,TRK,26.99


1. Remove Redundant Columns

*   The "Zero" Column: There is a column between product_id and product_quantity that contains only 0. If this represents a field like "Discount" that isn't being used, delete it to declutter the table.
*   Duplicate Revenue: total_revenue and calc_revenue appear to be identical. Choose one and remove the other to avoid confusion during analysis.

In [ ]:
# Check if the column 'product_id' contains only 0
if (orderlines_cl_df['product_id'] == 0).all():
    orderlines_cl_df.drop(columns=['product_id'], inplace=True)
    print("Column 'product_id' removed because it contained only zeros.")
else:
    print("Column 'product_id' contains other values and was kept.")

Column 'product_id' removed because it contained only zeros.


In [ ]:
orderlines_cl_df = orderlines_cl_df.drop(columns=['id'], errors='ignore')

In [ ]:
orderlines_cl_df.head()

,id_order,product_quantity,sku,unit_price,date,total_revenue,category,calc_revenue
0,299539,1,OTT0133,18.99,2017-01-01 00:07:19,18.99,OTT,18.99
1,299540,1,LGE0043,399.00,2017-01-01 00:19:45,399.00,LGE,399.00
2,299541,1,PAR0071,474.05,2017-01-01 00:20:57,474.05,PAR,474.05
3,299542,1,WDT0315,68.39,2017-01-01 00:51:40,68.39,WDT,68.39
4,299543,1,JBL0104,23.74,2017-01-01 01:06:38,23.74,JBL,23.74


Final step then download Orderlines cleaned 2

In [ ]:
orderlines_cl_df.to_csv('orderlines_cleaned_2.csv', index=False)
from google.colab import files
files.download('orderlines_cleaned_2.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Orders next**

In [ ]:
# your code here
orders_cl_df.sample(5)

,order_id,created_date,total_paid,state
61724,361265,2017-06-01 14:49:07,339.0,shopping basket
171012,471384,2017-12-27 19:50:19,818.91,pending
115260,414846,2017-10-20 09:22:48,0.0,shopping basket
112055,411641,2017-10-11 23:31:06,49.58,shopping basket
220232,520723,2018-03-03 23:59:04,2615.99,place order


the main optimization potential lies in Memory Usage and Data Types. Currently, 3 out of 4 columns are stored as object (strings), which is slow and memory-intensive.

Here is how you can optimize it:

*   **created_date** to Datetime: Converting this allows for easy time-based filtering and grouping (e.g., by month) while saving space.
*   **total_paid** to Float: This column likely contains currency values. You cannot perform math on strings; convert it to float64.
*   **state** to Category: Status columns (like "shipped", "cancelled") usually have many repeating strings. Converting them to category can reduce the memory footprint by up to 90% for that column.
*   **order_id** Downcasting: Since it's currently int64, you can downcast it to int32 to save space if the IDs aren't extremely large.



In [ ]:
# Convert to datetime
orders_cl_df['created_date'] = pd.to_datetime(orders_cl_df['created_date'])

# Convert to float (handles commas if necessary)
orders_cl_df['total_paid'] = pd.to_numeric(orders_cl_df['total_paid'].replace('[^0-9.-]', '', regex=True), errors='coerce')

# Convert to category
orders_cl_df['state'] = orders_cl_df['state'].astype('category')

# Downcast integer
orders_cl_df['order_id'] = pd.to_numeric(orders_cl_df['order_id'], downcast='integer')

Beyond the technical data types we already fixed, here are the strategic optimizations for your specific dataset (e-commerce orders):

1. Data Cleaning (The "Noise" Problem)

In 226k rows, many entries are likely abandoned carts (shopping basket). If you want to analyze actual revenue, these will skew your averages.

*   Optimization: Create a separate DataFrame for "Actual Sales" vs. "Potential Sales."
*   Check for Duplicates: Use df.duplicated(subset=['order_id']). If an order_id appears with both "shopping basket" and "completed," you might be double-counting transactions.


In [ ]:
#In dataset, shopping basket represents intent, while completed represents money. It is best to separate them.

# Remove duplicate orders (keep the most recent status)
orders_cl_df = orders_cl_df.sort_values('created_date').drop_duplicates('order_id', keep='last')

# Create a dedicated Revenue DataFrame
# Only include orders that are actually finished
confirmed_sales = orders_cl_df[orders_cl_df['state'] == 'completed'].copy()

# Filter out potential test orders or errors (e.g., total_paid = 0)
confirmed_sales = confirmed_sales[confirmed_sales['total_paid'] > 0]

2. Feature Engineering (Time Analysis)

Since you have a created_date, you can extract valuable columns to speed up future grouping:

*   Benefit: Instead of recalculating dates every time you plot a chart, you have "ready-to-use" categories for peak shopping hours or monthly growth.



In [ ]:
# Extract hour of the day (0-23)
orders_cl_df['hour'] = orders_cl_df['created_date'].dt.hour

# Extract day name for weekly trends
orders_cl_df['day_of_week'] = orders_cl_df['created_date'].dt.day_name()

# Extract month-year period for growth analysis
orders_cl_df['month_year'] = orders_cl_df['created_date'].dt.to_period('M')

In [ ]:
# 1. Convert Day of Week to Category (Massive memory saving)
# Define the order so your charts don't show "Friday" before "Monday"
days_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
orders_cl_df['day_of_week'] = pd.Categorical(
    orders_cl_df['day_of_week'],
    categories=days_order,
    ordered=True
)

# 2. Downcast Hour (Optional, just to be super clean)
orders_cl_df['hour'] = pd.to_numeric(orders_cl_df['hour'], downcast='unsigned')

# 3. Keep month_year as is, it's great for time series!

3. Handling Outliers in total_paid

E-commerce exports often contain test orders (0.00) or massive errors (99,999.00).


*   Optimization: Filter or flag these before doing a .mean() or .sum().


In [ ]:
# Identify potential test orders (0.00 or less)
test_orders = orders_cl_df[orders_cl_df['total_paid'] <= 0]

# Filter for realistic orders (e.g., between 1.00 and 20,000.00)
# Adjust the upper limit based on your actual high-ticket items
orders_cl_df = orders_cl_df[(orders_cl_df['total_paid'] > 0) & (orders_cl_df['total_paid'] < 20000)]

4. Memory Footprint (String Consolidation)

The state column has long strings like "shopping basket."

*   Optimization: If you only care about a few states, map them to shorter codes or integers to save even more memory, though category (which we did earlier) is usually sufficient.

In [ ]:
#Although category is good, mapping to short codes saves even more memory and makes queries faster.
# Map long strings to short, memory-efficient codes
state_map = {
    'shopping basket': 'basket',
    'place order': 'placed',
    'completed': 'done',
    'cancelled': 'void'
}

# Apply mapping and ensure it remains a category
orders_cl_df['state'] = orders_cl_df['state'].map(state_map).astype('category')

In [ ]:
orders_cl_df.sample(5)

,order_id,created_date,total_paid,state,hour,day_of_week,month_year
143022,443115,2017-11-25 12:32:24,263.23,placed,12,Saturday,2017-11
136116,435809,2017-11-23 20:41:13,335.74,basket,20,Thursday,2017-11
138940,438735,2017-11-24 12:19:42,78.47,basket,12,Friday,2017-11
80770,380320,2017-07-22 03:08:31,1854.98,void,3,Saturday,2017-07
185479,485887,2018-01-08 20:27:27,99.98,basket,20,Monday,2018-01


1. The "Zero-Value" Filter

In your sample, several rows (e.g., index 83951, 58186, 142544) have total_paid = 0.0. These are either failed attempts, test orders, or system errors that will ruin your "Average Order Value" (AOV) calculations.


In [ ]:
# 1. Convert to Numeric FIRST (Fixes the TypeError: str vs int)
# errors='coerce' turns non-numeric text into NaN so the code won't crash
orders_cl_df['total_paid'] = pd.to_numeric(orders_cl_df['total_paid'], errors='coerce')

# 2. Filter out Zero-values and NaNs (Cleans the data for revenue analysis)
orders_cl_df = orders_cl_df[orders_cl_df['total_paid'] > 0].dropna(subset=['total_paid']).copy()

# 3. Optimize other Data Types (Reduces memory usage)
orders_cl_df['created_date'] = pd.to_datetime(orders_cl_df['created_date'])
orders_cl_df['state'] = orders_cl_df['state'].astype('category')
orders_cl_df['order_id'] = pd.to_numeric(orders_cl_df['order_id'], downcast='integer')

# 4. Clean the Index (Organizes the rows from 0 to N)
orders_cl_df = orders_cl_df.reset_index(drop=True)

# Verify results
print(orders_cl_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207506 entries, 0 to 207505
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      207506 non-null  int32         
 1   created_date  207506 non-null  datetime64[ns]
 2   total_paid    207506 non-null  float64       
 3   state         193138 non-null  category      
 4   hour          207506 non-null  uint8         
 5   day_of_week   207506 non-null  category      
 6   month_year    207506 non-null  period[M]     
dtypes: category(2), datetime64[ns](1), float64(1), int32(1), period[M](1), uint8(1)
memory usage: 6.1 MB
None


2. Status Consolidation (The "Funnel" Logic)

You have pending, place order, and shopping basket. To make your analysis cleaner, you should group these into "In Progress" vs. "Completed".

In [ ]:
# 2. Status Harmonization
# 'place order' and 'pending' often mean the same thing in different systems
# Let's group them to make charts more readable
state_cleaning = {
    'place order': 'pending',
    'shopping basket': 'basket',
    'completed': 'completed',
    'pending': 'pending'
}
orders_cl_df['state'] = orders_cl_df['state'].map(state_cleaning).astype('category')

3. Reset Index

The index in your sample is fragmented (148197, 167720, etc.). This happens after filtering or sorting. Resetting it improves look-up speed and keeps the DataFrame clean.

In [ ]:
# 3. Handling the Index
# Drop the old fragmented index and create a clean 0 to N sequence
orders_cl_df = orders_cl_df.reset_index(drop=True)

In [ ]:
# 4. Final Sanity Check: Identify High-Value Customers
# Since you have high values (e.g., 4717.59), let's tag them
orders_cl_df['is_high_value'] = orders_cl_df['total_paid'] > orders_cl_df['total_paid'].quantile(0.95)

In [ ]:
orders_cl_df.sample(5)

,order_id,created_date,total_paid,state,hour,day_of_week,month_year,is_high_value
69699,375168,2017-07-10 09:07:24,3116.58,NaN,9,Monday,2017-07,True
143819,456902,2017-12-07 21:29:55,2246.56,NaN,21,Thursday,2017-12,False
97308,405867,2017-09-28 15:20:00,143.09,NaN,15,Thursday,2017-09,False
155112,469783,2017-12-22 16:25:09,23.98,NaN,16,Friday,2017-12,False
42917,345937,2017-04-13 13:52:24,71.96,NaN,13,Thursday,2017-04,False


1. The NaN Category Trap

In your previous sample, the state column had NaN values. In a category type, NaN values are invisible during a .groupby(). If you calculate total revenue, these orders will simply vanish from the report.

*   Optimization: Explicitly label them as 'unknown'.


In [ ]:
# 1. Handle NaN in Categories (Don't let orders disappear)
if orders_cl_df['state'].isnull().any():
    # We must add the category before we can fill NaNs with it
    orders_cl_df['state'] = orders_cl_df['state'].cat.add_categories('unknown').fillna('unknown')

2. Float Downcasting

You are using float64. Unless you are calculating scientific precision to 15 decimal places, float32 is perfect for currency (up to ~7 digits) and will reduce the memory of that column by 50%.

In [ ]:
# 2. Memory: Downcast Float64 to Float32
orders_cl_df['total_paid'] = orders_cl_df['total_paid'].astype('float32')

3. Duplicate Order Logic (The "Status Trail")

In e-commerce, an order_id often appears twice: once as shopping basket and later as completed. If you sum total_paid now, you are double-counting the money.

*   Optimization: Keep only the latest status for each order_id.

In [ ]:
# 3. Accuracy: Remove Duplicate Order IDs (Keep the most recent status)
# Sort by date first to ensure 'keep=last' gets the final status (e.g., 'completed')
orders_cl_df = orders_cl_df.sort_values('created_date').drop_duplicates('order_id', keep='last')

In [ ]:
# 4. Cleanup: Reset index after dropping duplicates
orders_cl_df = orders_cl_df.reset_index(drop=True)

In [ ]:
# Final check
print(orders_cl_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207506 entries, 0 to 207505
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   order_id       207506 non-null  int32         
 1   created_date   207506 non-null  datetime64[ns]
 2   total_paid     207506 non-null  float32       
 3   state          207506 non-null  category      
 4   hour           207506 non-null  uint8         
 5   day_of_week    207506 non-null  category      
 6   month_year     207506 non-null  period[M]     
 7   is_high_value  207506 non-null  bool          
dtypes: bool(1), category(2), datetime64[ns](1), float32(1), int32(1), period[M](1), uint8(1)
memory usage: 5.5 MB
None


In [ ]:
orders_cl_df.sample(5)

,order_id,created_date,total_paid,state,hour,day_of_week,month_year,is_high_value
20748,322183,2017-02-10 11:01:06,99.980003,unknown,11,Friday,2017-02,False
70751,376313,2017-07-12 11:31:21,212.990005,unknown,11,Wednesday,2017-07,False
40355,318032,2017-04-04 15:41:08,103.980003,unknown,15,Tuesday,2017-04,False
102734,411894,2017-10-12 21:10:40,88.980003,unknown,21,Thursday,2017-10,False
156895,471819,2017-12-26 08:34:00,65.480003,unknown,8,Tuesday,2017-12,False


1. The Floating-Point "Precision" Clean-up

Look at total_paid: values like 295.980011 or 880.989990 are artifacts of float32 conversion. In e-commerce, you want clean currency decimals.

*   Optimization: Round to 2 decimal places. This makes your reports look professional and prevents tiny summation errors.


In [ ]:
# 1. Round currency to 2 decimals (Fixes the .000001 artifact)
orders_cl_df['total_paid'] = orders_cl_df['total_paid'].round(2)

2. Time-Based Indexing (Optional but powerful)

Since you have 207k rows of time-series data, you can set the index to created_date.

*   Benefit: This allows for incredibly fast time-slices, like orders_cl_df['2017-12'] to instantly get all December orders.


In [ ]:
# 2. Sort by date for logical flow
orders_cl_df = orders_cl_df.sort_values('created_date')

3. State "Mapping" for Clean Funnels

You have basket, pending, completed, and unknown. If you are building a dashboard, "pending" and "basket" are often grouped as "unpaid" to see the conversion gap.


In [ ]:
# 3. Create a 'revenue_confirmed' flag
# This makes it easy to sum actual money without filtering every time
orders_cl_df['is_revenue'] = orders_cl_df['state'] == 'completed'

In [ ]:
# 4. Memory: If you don't need 'is_high_value' as an object,
# ensure it's a tight boolean
orders_cl_df['is_high_value'] = orders_cl_df['is_high_value'].astype(bool)

In [ ]:
# 5. Clean Index
orders_cl_df = orders_cl_df.reset_index(drop=True)

print(orders_cl_df.head())
print(orders_cl_df.info())

   order_id        created_date  total_paid    state  hour day_of_week  \
0    299539 2017-01-01 00:07:19   18.990000  unknown     0      Sunday   
1    299540 2017-01-01 00:19:45  399.000000  unknown     0      Sunday   
2    299541 2017-01-01 00:20:57  474.049988  unknown     0      Sunday   
3    299542 2017-01-01 00:51:40   68.389999  unknown     0      Sunday   
4    299543 2017-01-01 01:06:38   23.740000  unknown     1      Sunday   

  month_year  is_high_value  is_revenue  
0    2017-01          False       False  
1    2017-01          False       False  
2    2017-01          False       False  
3    2017-01          False       False  
4    2017-01          False       False  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207506 entries, 0 to 207505
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   order_id       207506 non-null  int32         
 1   created_date   207506 non-null  

In [ ]:
orders_cl_df.sample(5)

,order_id,created_date,total_paid,state,hour,day_of_week,month_year,is_high_value,is_revenue
86261,393547,2017-08-28 17:34:33,32.980000,unknown,17,Monday,2017-08,False,False
14979,315912,2017-01-27 15:04:59,186.979996,unknown,15,Friday,2017-01,False,False
89134,396746,2017-09-05 14:22:32,12.990000,unknown,14,Tuesday,2017-09,False,False
85659,392854,2017-08-26 12:38:28,178.979996,unknown,12,Saturday,2017-08,False,False
99632,408330,2017-10-04 17:06:02,1256.979980,unknown,17,Wednesday,2017-10,False,False


In [ ]:
# 1. Clean up floating-point artifacts (Professional currency look: 474.05 instead of 474.049988)
orders_cl_df['total_paid'] = orders_cl_df['total_paid'].round(2)

# 2. Data Integrity: Remove duplicate orders (Keep the latest status)
# In e-commerce, one order_id often has multiple rows (basket -> pending -> completed).
# Sorting by date and dropping duplicates ensures you don't triple-count revenue.
orders_cl_df = orders_cl_df.sort_values('created_date').drop_duplicates('order_id', keep='last')

# 3. Final Memory Check: Ensure Booleans are tight (already done, but good to keep)
orders_cl_df['is_high_value'] = orders_cl_df['is_high_value'].astype(bool)
orders_cl_df['is_revenue'] = orders_cl_df['is_revenue'].astype(bool)

# 4. Reset index for a clean sequence 0 to N
orders_cl_df = orders_cl_df.reset_index(drop=True)

#Keep the fast float32 data, but tell Pandas to only show two decimals:
pd.options.display.float_format = '{:,.2f}'.format
print(orders_cl_df.head())

# Verify the final state
print(orders_cl_df.info())
print(orders_cl_df.head())


   order_id        created_date  total_paid    state  hour day_of_week  \
0    299539 2017-01-01 00:07:19       18.99  unknown     0      Sunday   
1    299540 2017-01-01 00:19:45      399.00  unknown     0      Sunday   
2    299541 2017-01-01 00:20:57      474.05  unknown     0      Sunday   
3    299542 2017-01-01 00:51:40       68.39  unknown     0      Sunday   
4    299543 2017-01-01 01:06:38       23.74  unknown     1      Sunday   

  month_year  is_high_value  is_revenue  
0    2017-01          False       False  
1    2017-01          False       False  
2    2017-01          False       False  
3    2017-01          False       False  
4    2017-01          False       False  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207506 entries, 0 to 207505
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   order_id       207506 non-null  int32         
 1   created_date   207506 non-null  

In [ ]:
#Return to Float64 (Higher Precision)
#If the visual artifacts bother you during calculation, switch back to float64, which hides these approximations much better:
#orders_cl_df['total_paid'] = orders_cl_df['total_paid'].astype('float64').round(2)

In [ ]:
# Drop redundant index column
orders_cl_df = orders_cl_df.drop(columns=['is_high_value', 'is_revenue'])

KeyError: "['is_high_value', 'is_revenue'] not found in axis"

In [ ]:
# if state just unknown then drop
if (orders_cl_df['state'] == 'unknown').all():
    orders_cl_df.drop(columns=['state'], inplace=True)

In [ ]:
orders_cl_df.sample(100)

,order_id,created_date,total_paid,hour,day_of_week,month_year
162472,477866,2017-12-31 14:07:59,91.15,14,Sunday,2017-12
164745,480615,2018-01-03 10:01:24,103.98,10,Wednesday,2018-01
66784,371951,2017-07-03 14:26:16,"2,099.00",14,Monday,2017-07
169316,485728,2018-01-08 18:40:23,314.97,18,Monday,2018-01
90857,398700,2017-09-11 09:56:37,43.98,9,Monday,2017-09
...,...,...,...,...,...,...
6274,306489,2017-01-10 13:38:03,7.99,13,Tuesday,2017-01
171768,488453,2018-01-11 11:14:06,19.99,11,Thursday,2018-01
6062,306261,2017-01-10 08:21:07,326.97,8,Tuesday,2017-01
29353,331412,2017-03-02 20:15:33,9.99,20,Thursday,2017-03


In [ ]:
#show all variations and identify every unique formatting pattern in poc_2_df,
#you can use a "Structural Pattern Analysis."
#This converts digits into a placeholder (like 0) so you can see exactly where dots and lengths differ.

import re
import pandas as pd

def get_structure(val):
    # Replaces every digit with '0' to reveal the pattern (e.g., 1.234,56 -> 0.000,00)
    return re.sub(r'\d', '0', str(val))

# Define the columns to analyze
columns = ['order_id', 'created_date', 'total_paid', 'hour', 'day_of_week', 'month_year']

for col in columns:
    print(f"\n--- Pattern Analysis for: {col} ---")

    # Create a summary of patterns, their frequency, and unique examples
    analysis = (orders_cl_df[[col]].copy()
                .assign(pattern=orders_cl_df[col].apply(get_structure))
                .groupby('pattern')[col]
                .agg(['count', 'unique'])
                .sort_values(by='count', ascending=False))

    # Display the result
    print(analysis)


--- Pattern Analysis for: order_id ---
          count                                             unique
pattern                                                           
000000   207506  [299539, 299540, 299541, 299542, 299543, 29954...

--- Pattern Analysis for: created_date ---
                      count                                             unique
pattern                                                                       
0000-00-00 00:00:00  207506  [2017-01-01 00:07:19, 2017-01-01 00:19:45, 201...

--- Pattern Analysis for: total_paid ---
                     count                                             unique
pattern                                                                      
00.000000000000000   31073  [18.99, 23.74, 23.98, 24.99, 28.49, 19.99, 41....
000.00000000000000   27905  [474.05, 149.14, 225.98, 157.18, 154.48, 219.4...
00.00000000000000    26708  [68.39, 51.48, 72.19, 36.98, 59.84, 65.89, 35....
000.0000000000000    24466  [175.7, 263.48, 17

/tmp/ipykernel_567/918584434.py:21: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  .groupby('pattern')[col]


          count                                             unique
pattern                                                           
0000-00  207506  [2017-01, 2017-02, 2017-03, 2017-04, 2017-05, ...


In [ ]:
# 1. Round to 2 decimal places to remove the floating-point "noise"
orders_cl_df['total_paid'] = orders_cl_df['total_paid'].round(2)

# 2. If you want to FORCE the display to always show 2 decimals in your IDE/Notebook:
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format

# 3. Verify the patterns again
print(orders_cl_df['total_paid'].head())

0    18.99
1   399.00
2   474.05
3    68.39
4    23.74
Name: total_paid, dtype: float32


Final step then Download Orders cleaned 2

In [ ]:
orders_cl_df.to_csv('orders_cleaned.csv', index=False)
from google.colab import files
files.download('orders_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## 3.&nbsp; Exclude orders with unknown products


In [ ]:
# your code here
products_cl_df.sample(5)

,sku,name,desc,price,promo_price,in_stock,type,price_cleaned,promo_price_cleaned,price_ratio
4826,OTT0141,otterbox clearly protected gel iphone case se ...,a transparent cover piece made of flexible tpu...,19.99,129.906,False,11865403,19.99,12.99,0.65
9043,SYN0184,synology ram 4gb 1866mhz ddr3,4gb ddr3 memory for synology.,75.99,749.898,True,1404,75.99,74.99,0.99
4353,TUC0271,"angolo tucano ipad case pro 97 ""red",flap case for ipad camera pro 97,34.99,249.901,False,12635403,34.99,24.99,0.71
8024,APP2116,apple nike sport strap 42mm obsidian / black,durable and flexible strap sports with clip cl...,59,569.995,True,2449,59.00,57.00,0.97
1739,SEA0073,seagate backup plus slim external hard drive 2...,lightweight external hard drive and 25-inch ul...,110,989.901,False,11935397,110.00,98.99,0.90


In [ ]:
products_cl_1_df = products_cl_df.copy()

Recommended Data Structure

Keep only these columns for a clean analysis:

*   sku: Unique product identifier.
*   name: Product title.
*   desc: Product description.
*   in_stock: Availability status.
*   price_cleaned: The corrected base price.
*   promo_price_cleaned: The corrected promotional price.


In [ ]:
# Drop useless columns
products_cl_1_df = products_cl_1_df.drop(columns=['price', 'promo_price', 'price_ratio'])

In [ ]:
# List the columns you want to REMOVE
#cols_to_drop = ['price', 'promo_price', 'price_ratio']

# Drop the columns from your dataframe
#products_cl_1_df_cleaned = products_cl_1_df.drop(columns=cols_to_drop)

# Optional: Also remove rows where the product name is "Unknown" or missing
products_cl_1_df = products_cl_1_df.dropna(subset=['name', 'sku'])

print(products_cl_1_df.head())

       sku                                           name  \
0  RAI0007              silver rain design mstand support   
1  APP0023              apple mac keyboard keypad spanish   
2  APP0025               mighty mouse apple mouse for mac   
3  APP0072  apple dock to usb cable iphone and ipod white   
4  KIN0007    mac memory kingston 2gb 667mhz ddr2 so-dimm   

                                                desc  in_stock      type  \
0       aluminum support compatible with all macbook      True      8696   
1          usb ultrathin keyboard apple mac spanish.     False  13855401   
2                             mouse apple usb cable.     False      1387   
3              iphone dock and usb cable apple ipod.     False      1230   
4  2gb ram mac mini and imac (2006/07) macbook pr...      True      1364   

   price_cleaned  promo_price_cleaned  
0          59.99                49.99  
1          59.00                59.00  
2          59.00                56.99  
3          25.00

In [ ]:
# 1. Store the original count
original_count = products_cl_1_df.shape[0]

# 2. Drop the duplicates
cols = ['sku', 'name', 'desc', 'in_stock', 'type', 'price_cleaned', 'promo_price_cleaned']
# Keep the first occurrence and delete the rest
products_cl_1_df = products_cl_1_df.drop_duplicates(subset=cols, keep='first')

# 3. Store the new count
new_count = products_cl_1_df.shape[0]

# 4. Display the results
print(f"Rows before: {original_count}")
print(f"Rows after:  {new_count}")
print(f"Removed:     {original_count - new_count}")

Rows before: 10580
Rows after:  10580
Removed:     0


In [ ]:
products_cl_1_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10580 entries, 0 to 10579
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   sku                  10580 non-null  object 
 1   name                 10580 non-null  object 
 2   desc                 10580 non-null  object 
 3   in_stock             10580 non-null  bool   
 4   type                 10530 non-null  object 
 5   price_cleaned        10580 non-null  float64
 6   promo_price_cleaned  10580 non-null  float64
dtypes: bool(1), float64(2), object(4)
memory usage: 506.4+ KB


In [ ]:
# Display only the rows where 'type' is missing
missing_type_rows = products_cl_1_df[products_cl_1_df['type'].isna()]
print(missing_type_rows)

# Or simply count them
print(products_cl_1_df.isnull().sum())

             sku                                               name  \
147      SAN0017       sandisk cruzer edge usb 2.0 flash drive 16gb   
370      SAN0026    sandisk extreme cruzer 16gb usb 3.0 flash drive   
596      APP0672                    apple iphone 5s 16gb space gray   
601      APP0673                        apple iphone 5s 16gb silver   
602      APP0675                    apple iphone 5s 32gb space gray   
996      APP0823                         apple iphone 6 16gb silver   
1002     APP0829                    apple iphone 6 plus 16gb silver   
1003     APP0822                     apple iphone 6 16gb space gray   
1004     APP0825                     apple iphone 6 64gb space gray   
1005     APP0826                         apple iphone 6 64gb silver   
1006     APP0828                apple iphone 6 plus 16gb space gray   
1083     APP0856                          apple iphone 6 128gb gold   
1084     APP0857                apple iphone 6 plus 64gb space gray   
1085  

In [ ]:
products_cl_1_df.sample(3)

,sku,name,desc,in_stock,type,price_cleaned,promo_price_cleaned
4642,WDT0231-A,open - western digital elements 2tb usb 3.0 ha...,2tb portable external hard drive for mac and pc,False,1298,105.00,94.16
4190,OWC0172,6g ssd expansion of 480gb owc aura pro retina ...,expansion of 480gb ssd for macbook pro retina ...,False,12215397,398.99,306.58
5132,IOT0022,iottie easy one touch mini car iphone support ...,mini support for car vent all models of iphone,True,5720,29.99,20.99


1. Fix the type Column (Scientific Notation)

The type column is currently an object (string/mixed), which is why you see values like 2,17E+11. This happens when large IDs are treated as numbers. Converting them to Strings immediately prevents further corruption.

2. Handle Missing Values in type

You have 50 missing values in the type column (10,580 total vs 10,530 non-null). You should fill these to avoid errors during filtering.

In [ ]:
# 1. Clean the 'type' column
# 2. Convert to string to fix scientific notation (e.g., 5.74E+15) and fill NaNs
products_cl_1_df['type'] = products_cl_1_df['type'].fillna('Unknown').astype(str)

3. Exclude "Unknown" Products

Based on your original goal, we should filter out products that are clearly test entries or placeholders (e.g., names containing "test", "unknown", or "delete").

In [ ]:
# 3. Exclude "Unknown" or Placeholder Products
# This filters out rows where SKU or Name contains test keywords
exclude_mask = (
    products_cl_1_df['name'].str.contains('test|unknown|dummy|delete', case=False, na=False) |
    products_cl_1_df['sku'].str.contains('test|unknown|dummy|delete', case=False, na=False)
)

In [ ]:
# Apply filter and create a fresh copy to save memory
products_cl_1_df = products_cl_1_df[~exclude_mask].copy()

# 3. Memory & Type Optimization
products_cl_1_df['sku'] = products_cl_1_df['sku'].astype(str)
products_cl_1_df['name'] = products_cl_1_df['name'].astype(str)

# 4. (Optional) Add a 'discount' column for better analysis
products_cl_1_df['discount'] = products_cl_1_df['price_cleaned'] - products_cl_1_df['promo_price_cleaned']

# Final check
print(products_cl_1_df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 10578 entries, 0 to 10579
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   sku                  10578 non-null  object 
 1   name                 10578 non-null  object 
 2   desc                 10578 non-null  object 
 3   in_stock             10578 non-null  bool   
 4   type                 10578 non-null  object 
 5   price_cleaned        10578 non-null  float64
 6   promo_price_cleaned  10578 non-null  float64
 7   discount             10578 non-null  float64
dtypes: bool(1), float64(3), object(4)
memory usage: 671.5+ KB
None


Yes, it looks like the code was successfully implemented.

Here is the evidence from your info() output:

*   Exclusion Worked: Your row count dropped from 10,580 to 10,578. Two "unknown" or "test" rows were correctly identified and removed. [1, 2]
*   Missing Values Fixed: The type column now has 10,578 non-null entries (matching the total row count). All 50 missing values were filled. [3]
*   New Feature Added: The discount column (float64) has been added to the end of the DataFrame as requested. [4, 5]

**Next Step**

1. Optimize Memory (Downcasting)

Your memory usage is 671.5 KB. By converting generic "objects" into "categories" (for repeating values like type) and downcasting floats, you can reduce this significantly.


In [ ]:
# Convert 'type' to category (saves memory for repeating strings)
products_cl_1_df['type'] = products_cl_1_df['type'].astype('category')

# Downcast floats (from float64 to float32)
cols = ['price_cleaned', 'promo_price_cleaned', 'discount']
products_cl_1_df[cols] = products_cl_1_df[cols].apply(pd.to_numeric, downcast='float')

print(f"New Memory Usage: {products_cl_1_df.memory_usage(deep=True).sum() / 1024:.2f} KB")

New Memory Usage: 3134.48 KB


2. Scientific Notation Cleanup (Visual)

Even though the data is "cleaned," the type column might still show strings like 5.74E+15 because they were imported as floats. You can force them into a clean integer-string format:

In [ ]:
# Remove decimals and scientific notation from the 'type' string
products_cl_1_df['type'] = products_cl_1_df['type'].str.replace(r'\.0$', '', regex=True)

3. Consistency Filter: Negative Discounts

Sometimes data errors lead to a promo_price being higher than the price. You should check for and handle these "negative discounts":

In [ ]:
# Check for price errors
price_errors = products_cl_1_df[products_cl_1_df['discount'] < 0]

# Option: Set discount to 0 if promo_price was accidentally higher than original
products_cl_1_df.loc[products_cl_1_df['discount'] < 0, 'discount'] = 0

4. Create a "Brand" Column

Since your sku follows a pattern (e.g., APP for Apple, LAC for LaCie), you can extract the Brand for better grouping:

In [ ]:
# Extract first 3 letters of SKU as 'brand'
products_cl_1_df['brand'] = products_cl_1_df['sku'].str[:3].str.upper()

Analysis for Brand e.g.

To see which brand offers the most savings, we first extract the brand code from the sku and then aggregate the discount values.

In [ ]:
# 1. Extract the Brand (first 3 letters of SKU)
products_cl_1_df['brand'] = products_cl_1_df['sku'].str[:3].str.upper()

# 2. Create the Summary Table
brand_summary = products_cl_1_df.groupby('brand').agg(
    total_discount=('discount', 'sum'),
    avg_discount=('discount', 'mean'),
    product_count=('sku', 'count')
).sort_values(by='total_discount', ascending=False)

# Display the Top 5 Brands by Total Discount
print(brand_summary.head(5))

       total_discount  avg_discount  product_count
brand                                             
PAC     477767.500000    300.671814           1589
APP     185454.875000    103.954529           1784
AP2      61329.257812    175.728531            349
GTE      29902.140625    257.777069            116
OWC      20744.960938     58.272362            356


In [ ]:
products_cl_1_df.sample(3)

,sku,name,desc,in_stock,type,price_cleaned,promo_price_cleaned,discount,brand
2051,DRB0012-A,(open) dr bott digital video link dvi to mini ...,dvi to mini displayport converter for 27-inch ...,False,1298,118.00,49.00,69.00,DRB
6504,ZAG0018-A,"open - zagg ipad keyboard pro messenger 129 ""b...",cover with ultra resistant reconditioned and b...,False,54025401,89.99,23.60,66.39,ZAG
6757,ELA0015,elago space gray mouse skin,mat elegant original leather with reinforced e...,True,1387,45.00,32.99,12.01,ELA


In [ ]:
# A. Products with a price of 0.00 (Highly unlikely for electronics)
zero_prices = products_cl_1_df[products_cl_1_df['price_cleaned'] <= 0]

# B. Negative Discounts (Promo price is higher than the original price)
negative_discounts = products_cl_1_df[products_cl_1_df['promo_price_cleaned'] > products_cl_1_df['price_cleaned']]

# C. Extreme outliers (Anything over 20,000 € is usually a data entry error in this catalog)
extreme_outliers = products_cl_1_df[products_cl_1_df['promo_price_cleaned'] > 20000]

print("--- Zero Price Items ---")
print(zero_prices[['sku', 'name', 'price_cleaned']])

print("\n--- Items where Promo > Original ---")
print(negative_discounts[['sku', 'name', 'price_cleaned', 'promo_price_cleaned']])

print("\n--- Extreme Outliers (> 20k) ---")
print(extreme_outliers[['sku', 'name', 'promo_price_cleaned']])

--- Zero Price Items ---
            sku                                               name  \
34      TWS0019  twelve south magicwand support apple magic tra...   
1579    AII0008           aiino case macbook air 11 '' transparent   
1717    CEL0020  celly ambo luxury leather case + iphone 6 case...   
1720    CEL0007  celly wallet case with removable cover black i...   
1721    CEL0012            celly silicone hard shell iphone 6 blue   
1722    CEL0014        celly silicone hard shell iphone 6 amarillo   
1727    CEL0015    celly fur-lined powerbank battery 4000mah black   
1729    CEL0018     celly wallet leather case cover black iphone 6   
1730    CEL0023  celly ambo luxury leather case + case gold iph...   
1731    CEL0025  celly ambo luxury leather case + case iphone 6...   
2805    APP1197  apple imac 21.5 "core i5 31 ghz retina display...   
4293    CEL0028  celly swing arm button selfie with bluetooth b...   
4294    CEL0029  celly extensible arm selfie with bluetooth but..

In [ ]:
# Keep only products with a valid price and remove extreme errors
products_cl_1_df = products_cl_1_df[
    (products_cl_1_df['price_cleaned'] > 0) &
    (products_cl_1_df['promo_price_cleaned'] <= 20000) &
    (products_cl_1_df['promo_price_cleaned'] <= products_cl_1_df['price_cleaned'])
].copy()

# Round the discount to 2 decimals to fix floating point math
products_cl_1_df['discount'] = products_cl_1_df['discount'].round(2)

print(f"Remaining products after outlier removal: {len(products_cl_1_df)}")

Remaining products after outlier removal: 10532


To calculate the potential revenue loss (the total amount "given away" as discounts), we sum the discount column. However, it is vital to perform this calculation after removing the outliers (like the 240,000 € powerbank), or the result will be completely distorted.


In [ ]:
# 1. First, ensure outliers are gone (Price 0 and the 240k Powerbank)
products_cl_1_df = products_cl_1_df[
    (products_cl_1_df['price_cleaned'] > 0) &
    (products_cl_1_df['promo_price_cleaned'] < 20000)
].copy()

# 2. Calculate Total Potential Revenue Loss
total_potential_loss = products_cl_1_df['discount'].sum()

# 3. Calculate Average Discount per Product
avg_discount_per_item = products_cl_1_df['discount'].mean()

print(f"Total Potential Revenue Loss: {total_potential_loss:,.2f} €")
print(f"Average Discount per Product: {avg_discount_per_item:.2f} €")

Total Potential Revenue Loss: 1,031,883.31 €
Average Discount per Product: 97.98 €


In [ ]:
Final step then download products cleaned 2

In [ ]:
products_cl_1_df.to_csv('products_cleaned_2.csv', index=False)
from google.colab import files
files.download('products_cleaned_2.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## 4.&nbsp; Explore the revenue from different tables

#### Step 1:
Create the `unit_price_total` as `orderlines.unit_price` * `orderlines.product_quantity`

In [ ]:
# your code here

#### Step 2:
Group by `id_order`, summarising by the sum of `unit_price_total`

In [ ]:
# your code here

### What is the average difference between `total_paid` and `unit_price_total`?

In [ ]:
# your code here

### What is the distribution of these differences?

In [ ]:
# your code here

### Can all the differences be explained by shipping costs? If not, what are other plausible explanations?

In [ ]:
# your explanation here

### If there are differences that you can’t explain: what should you do with these orders?

In [ ]:
# your code here

## 5.&nbsp; Become confident about your dataset

Let's hear your thoughts about what you discovered.

In [ ]:
# your thoughts & code here

Do not forget to download your quality controlled DataFrames

In [ ]:
from google.colab import files

#orders_quality_controlled.to_csv("orders_qu.csv", index=False)
#files.download("orders_qu.csv")

#orderlines_quality_controlled.to_csv("orderlines_qu.csv", index=False)
#files.download("orderlines_qu.csv")